feat: Begin transactions before each transaction read/write

danieljbruce opened the pull request:
**Summary**

This PR ensures that transactions are always started before the first transaction read/write call. It also solves various concurrency problems with transactions like ensuring that two transactions are never started at the same time.

**Source code changes:**

A middleware function called withBeginTransaction is added that will accept custom code as a parameter and ensure a transaction has been started before running that custom code. This function also ensures that two transactions are never started at the same time.

Each read function including get, runQuery and runAggregationQuery as well as the commit function will now use withBeginTransaction to ensure a transaction has been started before doing what these read and write functions normally do.

The run function is also modified so that two transactions are never started at the same time. The modifications to the run function also ensure a transaction object cannot be used to start a transaction a second time.

**Test changes:**

System tests are added to capture various code samples where a transaction is used. Each test calls various transaction functions in a particular order with and without using the run function first. Each test ensures transactions behave properly. A few other existing tests are modified so that errors will bubble up to the test runner for easier debugging.

Some changes are made to unit tests from the previous PR. First, any time a request is used it is constrained to a particular type that must match a Gapic read/write type. This enforces stronger type checking in the tests. Stronger type checking for other variables is added as well. An enum was added and each value of the enum matches a function name from the gap layer. Values for this enum are used when mocking out functions instead of strings in order to eliminate magic string matching in the code.

New unit tests are added that test concurrency when using transactions. This new test suite defines expected responses, expected requests and defines a new object called TransactionOrderTester. The TransactionOrderTester is used by the test to ensure events in code occur in a particular order. The TransactionOrderTester is also used to ensure that code execution events as well requests occur in a particular order specified by each test.

danieljbruce committed:
Add a unit test for commit as a non-tx

The fact that using a transaction object to do a commit results in a non-transaction should be documented so that if we decide to introduce a change later where this behaves differently then it is well documented.

# Conflicts:
#	test/transaction.ts

danieljbruce committed:
use linter

danieljbruce committed:
Write tests to capture current behavior of error

When begin transaction sends back an error, we want some tests to capture what the behavior is so that when we make changes to the run function then behavior is preserved.

danieljbruce committed:
Add tests for passing a response

A response should reach the user the right way. Add tests to make sure behavior is preserved.

danieljbruce committed:
run async close to working

In the run function delegate calls to runAsync and use run async to make promise calls

danieljbruce committed:
Add runAsync to promise excludes

This allows this function to return a promise instead of a promise wrapped in a promise. This makes the tests pass and behave the way they should.

danieljbruce committed:
Remove space

danieljbruce committed:
Change to use this instead of self

Do not call request from self

danieljbruce committed:
Eliminate unused comments

danieljbruce committed:
Add two comments

Comments should actually explain what is being done

danieljbruce committed:
Remove the commit test for this PR

The commit test for this PR should be removed because it is not really relevant for the async run functionality.

danieljbruce committed:
Clarify types throughout the function

The types used should be very specific so that reading the code isn’t confusing.

danieljbruce committed:
Add a bit more typing for clarity

Add a type to the resolve function just to introduce more clarity

danieljbruce committed:
Change types used in the data client callback

Make the types more specific in the data client callback so that it is easier to track down the signature and match against the begin transaction function.

danieljbruce committed:
run the linter

danieljbruce committed:
Add comments to clarify PR

danieljbruce committed:
Refactor the parsing logic out of run

The parsing logic is going to be needed elsewhere so taking it apart now.

danieljbruce committed:
Change interface of request promise callback

The interface name should be changed so that it matches what it is. It is the callback used to form a promise.

danieljbruce committed:
Move commit functionality to a new function

The internals of commit should be moved to a new function. This way we can sandwich a commit async call between commit and runCommitAsync.

danieljbruce committed:
Add the mutex and transaction state etc.

Still an issue with system tests. commit still needs tweaks to work with async.

danieljbruce committed:
Remove no-op, get commit tests working

Remove the no-op, get commit tests in place.

danieljbruce committed:
Add mocks and additional debugging

Mocks and additional debugging hooks to introspect what is going on.

danieljbruce committed:
Add tests for commit

Make sure commit behaves the same way as before.

danieljbruce committed:
Add the commit tests

Commit tests added to ensure that commit behaves the same way as before.

danieljbruce committed:
Fix the tests so that they pass on commit

The tests should pass before we make changes to commit.

danieljbruce committed:
refactor one of the mocks

One of the mocks does not need to be written twice

danieljbruce committed:
Merge branch 'nodejs-transaction-redesign-feature-branch-1a-add-commit-tests' of https://github.com/danieljbruce/nodejs-datastore into nodejs-transaction-redesign-feature-branch-2-adding-commit

danieljbruce committed:
reverting changes to add new test on transaction

danieljbruce committed:
Merge branch 'nodejs-transaction-redesign-feature-branch-1a-add-commit-tests' of https://github.com/danieljbruce/nodejs-datastore into nodejs-transaction-redesign-feature-branch-3-introducing-async-commit

# Conflicts:
#	test/transaction.ts

danieljbruce committed:
Change the promise

Make the promise simpler. Change the tests to exclude functions with promisify.

danieljbruce committed:
Hide data completely

Change accessors to hide data completely instead of using the private modifier

danieljbruce committed:
PR use if/else block

Eliminate the early return as suggested in the PR

danieljbruce committed:
Add comments to document the new functions

The comments capture the parameters and return type.

danieljbruce committed:
Update return type in docs

danieljbruce committed:
Update the tests to include runAsync

runAsync should be in promisfy excludes

danieljbruce committed:
refactor: Break transaction.run into smaller pieces for use with async functions and other read/write calls

danieljbruce committed:
Merge branch 'nodejs-transaction-change-run' of https://github.com/danieljbruce/nodejs-datastore into nodejs-transaction-change-run

danieljbruce committed:
Rename a function to be more descriptive

Make sure it is explicit that we are parsing begin results.

danieljbruce committed:
Move the mutex and the state down to derived class

The mutex and state should be moved down to the derived class. We are going to override get/runQuery/runAggregateQuery there.

renovate-bot committed:
chore(deps): update dependency @types/sinon to v17 (#1197)

danieljbruce committed:
Add hook to call run before commit

Add the hook to call run before calling commit in existing tests.

danieljbruce committed:
Add commitAsync to promisify excludes

commitAsync should be resolved and then() function should be called as it was not being called before

danieljbruce committed:
remove the console logs

danieljbruce committed:
Delete run commit

danieljbruce committed:
Remove the private identifier

Use the private modifier instead to hide data

danieljbruce committed:
Add withBeginTransaction

withBeginTransaction will be used with all calls that begin transactions and then intend to use the mutex for locking when the begin transaction call is made.

danieljbruce committed:
Add another level of abstraction

Add #beginWithCallback so that all the read calls can be made with one line of code.

danieljbruce committed:
commit async is not needed anymore

danieljbruce committed:
This data structure is not needed anymore

danieljbruce committed:
Replace types associated with run to use generics

The generic parameter should be used for types with run.

danieljbruce committed:
Make response type more specific for parseRunAsync

More specific types are better and it make code easier to read

renovate-bot committed:
chore(deps): update dependency @types/is to v0.0.25 (#1198)

[![Mend Renovate](https://app.renovatebot.com/images/banner.svg)](https://renovatebot.com)

This PR contains the following updates:

| Package | Change | Age | Adoption | Passing | Confidence |
|---|---|---|---|---|---|
| [@types/is](https://togithub.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/is) ([source](https://togithub.com/DefinitelyTyped/DefinitelyTyped)) | [`0.0.24` -> `0.0.25`](https://renovatebot.com/diffs/npm/@types%2fis/0.0.24/0.0.25) | [![age](https://developer.mend.io/api/mc/badges/age/npm/@types%2fis/0.0.25?slim=true)](https://docs.renovatebot.com/merge-confidence/) | [![adoption](https://developer.mend.io/api/mc/badges/adoption/npm/@types%2fis/0.0.25?slim=true)](https://docs.renovatebot.com/merge-confidence/) | [![passing](https://developer.mend.io/api/mc/badges/compatibility/npm/@types%2fis/0.0.24/0.0.25?slim=true)](https://docs.renovatebot.com/merge-confidence/) | [![confidence](https://developer.mend.io/api/mc/badges/confidence/npm/@types%2fis/0.0.24/0.0.25?slim=true)](https://docs.renovatebot.com/merge-confidence/) |

---

### Configuration

📅 **Schedule**: Branch creation - "after 9am and before 3pm" (UTC), Automerge - At any time (no schedule defined).

🚦 **Automerge**: Disabled by config. Please merge this manually once you are satisfied.

♻ **Rebasing**: Whenever PR is behind base branch, or you tick the rebase/retry checkbox.

🔕 **Ignore**: Close this PR and you won't be reminded about this update again.

---

 - [ ] <!-- rebase-check -->If you want to rebase/retry this PR, check this box

---

This PR has been generated by [Mend Renovate](https://www.mend.io/free-developer-tools/renovate/). View repository job log [here](https://developer.mend.io/github/googleapis/nodejs-datastore).
<!--renovate-debug:eyJjcmVhdGVkSW5WZXIiOiIzNy40Ni4wIiwidXBkYXRlZEluVmVyIjoiMzcuNDYuMCIsInRhcmdldEJyYW5jaCI6Im1haW4ifQ==-->

danieljbruce committed:
Modify comment

Modify comment so that it doesn’t reference the way the code was before.

danieljbruce committed:
Add implementation for runQuery

runQuery should make the call to begin the transaction first if that hasn’t already happened yet.

danieljbruce committed:
Making fixes for runAggregationQuery

Fixes for run aggregation query. Still getting undefined results.

danieljbruce committed:
Write tests for runAggregateQuery

Make sure that runAggregateQuery return results make it back to the user.

danieljbruce committed:
Get one test case passing for runAggregateQuery

runAggregateQuery should not be excluded by promisify. Otherwise it will not return a promise, but we want it to return a promise.

danieljbruce committed:
remove console log clutter

danieljbruce committed:
Change tests for runAggregationQuery

Change the test to use deep strict equal since the objects being compared will not be reference equal.

danieljbruce committed:
Add resolver type

Eliminate some unused code

danieljbruce committed:
Create a transaction wrapper class for testing

This is going to be a useful test for mocking out various layers to make sure they work the same way as before.

danieljbruce committed:
Clean up mocked transaction wrapper

The mocked transaction wrapper should reset all mocked gapic functions and not have to be told which ones to reset.

danieljbruce committed:
Move test information for commit into commit block

danieljbruce committed:
Add gapic mocked tests for aggregation query

Gapic mocked tests for run aggregation query need to be written that use the mock transaction object.

danieljbruce committed:
Fixing up the runQuery test

danieljbruce committed:
Finished the runQuery tests

The runQuery tests are finished so we can use this to take apart the function from end to end.

danieljbruce committed:
Finished the get tests

The get tests now make sure that the data coming back from the gapic layer results in the same values for users.

danieljbruce committed:
remove only

danieljbruce committed:
Merge branch 'nodejs-transaction-redesign-feature-branch-1a-add-commit-tests' of https://github.com/danieljbruce/nodejs-datastore into nodejs-transaction-redesign-feature-branch-4-all-call-types

# Conflicts:
#	test/transaction.ts

danieljbruce committed:
remove only and console log

danieljbruce committed:
Try modifications to runQuery

Add modifications to runQuery to use some function to return values

danieljbruce committed:
Add try catch blocks to handle errors in the tests

try/catch logic in the test suite is added so that errors bubble up to the test runner and it is easier to see why tests failed

danieljbruce committed:
Merge branch 'nodejs-transaction-redesign-feature-branch-1a-add-commit-tests' of https://github.com/danieljbruce/nodejs-datastore into nodejs-transaction-redesign-feature-branch-4-all-call-types

danieljbruce committed:
Modify commit so it doesn’t run early

Don’t start commit with a promise or the promise will run early.

danieljbruce committed:
Update get with resolver

get should use the same pattern as runQuery and runAggregationQuery to use a resolver for the mutex business logic

danieljbruce committed:
remove #beginWithCallback

#beginWithCallback is no longer used so remove it

danieljbruce committed:
Remove #withBeginTransaction

This function is no longer used

danieljbruce committed:
Rename #someFunction with begin transaction

#someFunction should be named differently

danieljbruce committed:
Fix promisify. Change to a deepStrictEqual check.

A deepStrictEqual check is all that is needed in this test that currently reuses a transaction.

danieljbruce committed:
Add setImmediate to the tests

If we add a delay to the tests then the mutex has the opportunity to unlock before running the tests.

danieljbruce committed:
Added some tests for call ordering

We want to make sure that the calls get delivered and received in order

danieljbruce committed:
Pack testing tool into an object

Put the testing tool in an object so that we can explore more orderings and possibilities.

danieljbruce committed:
Eliminate console logs, use expected order

Allow expected order to be passed into the tester. This will make the object reusable.

danieljbruce committed:
Add a check for transaction not started yet

Need a check to be sure that the transaction is not in progress if making another beginTransaction call.

danieljbruce committed:
Merge branch 'nodejs-transaction-change-run' of https://github.com/danieljbruce/nodejs-datastore into nodejs-transaction-redesign-feature-branch-1a-add-commit-tests

danieljbruce committed:
Merge branch 'nodejs-transaction-redesign-feature-branch-1a-add-commit-tests' of https://github.com/danieljbruce/nodejs-datastore into nodejs-transaction-redesign-feature-branch-2-adding-commit

# Conflicts:
#	src/transaction.ts

danieljbruce committed:
Merge branch 'nodejs-transaction-redesign-feature-branch-2-adding-commit' of https://github.com/danieljbruce/nodejs-datastore into nodejs-transaction-redesign-feature-branch-2a-adding-commit-and-tests

# Conflicts:
#	test/transaction.ts

danieljbruce committed:
Remove NOT_TRANSACTION

Remove the NOT_TRANSACTION and default to NOT_STARTED

danieljbruce committed:
Remove only

Remove only and let all unit tests run

danieljbruce committed:
Use an enum instead of static class members

Use an enum to track transaction state.

danieljbruce committed:
TODOs are done

danieljbruce committed:
Move excludes to proper position

Excludes should contain the right functions in the right order.

danieljbruce committed:
Simplify the run function a little bit

Regroup functionality to simplify the run function

danieljbruce committed:
Merge branch 'nodejs-transaction-redesign-feature-branch-2a-adding-commit-and-tests' of https://github.com/danieljbruce/nodejs-datastore into nodejs-transaction-redesign-feature-branch-4-all-call-types

# Conflicts:
#	src/transaction.ts
#	test/transaction.ts

danieljbruce committed:
Add comments to tests

Explain purpose of testing objects

danieljbruce committed:
Modify tests and fix a bug from the merge

The merge added a bug because the pound sign wasn’t used and it should have been.

danieljbruce committed:
Fix error message

danieljbruce committed:
Comments and general cleanup

Remove constructor parameter that is not used. Add comments to functions so that the tests are more readable.

danieljbruce committed:
Added comments for functions in transaction.ts

Functions in transaction.ts need JSdoc comments.

danieljbruce committed:
Add a comment for aggregate queries

Add JSdoc comment

danieljbruce committed:
Add an assertion to test

The assertion should make sure the first read returns undefined.

danieljbruce committed:
Add tests for lookup, put, commit

For calls with run and without run, do lookup, put, commit. Make sure to clean up after the tests too.

danieljbruce committed:
refactor the test

Test does lookup, put, commit. This is done in an async function.

danieljbruce committed:
Add tests for put, lookup, commit

A group of tests should be added for put, lookup, commit. One with run and one without run.

gcf-owl-bot[bot] committed:
chore: update cloud-rad version to ^0.4.0 (#1199)

Source-Link: https://github.com/googleapis/synthtool/commit/1063ef32bfe41b112bade7a2dfad4e84d0058ebd
Post-Processor: gcr.io/cloud-devrel-public-resources/owlbot-nodejs:latest@sha256:e92044720ab3cb6984a70b0c6001081204375959ba3599ef6c42dd99a7783a67

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

danieljbruce committed:
Add comments and general cleanup

Cleanup includes adding comments so that the tests are more readable.

danieljbruce committed:
Remove some redundant tests

Some of these tests are captured in the various functions describe block.

danieljbruce committed:
Correct the comment to be more accurate

The comment should talk about what the tests actually intend to do

danieljbruce committed:
General improvements to code quality

Move declared transactionWrapper and other variables out to shorten code. Remove require and replace with import. Use more specific type.

danieljbruce committed:
Add data client check

Add data client check and add type for data client.

danieljbruce committed:
Eliminate the mocked function variable

The mocked function variable is not used so remove it from code.

danieljbruce committed:
Move begin transaction setup code

The same block of code for setting up begin transaction is repeated twice. Move it into one function and use it from both before blocks.

danieljbruce committed:
Replace the TODO for the key

Add some try blocks to make the errors more visible also.

danieljbruce committed:
Update description

Update the description for the test to give a better explanation of what the describe block does.

danieljbruce committed:
Add comments to describe the purpose of signaller

The comments describing the callback signaller should explain the problem it solves.

danieljbruce committed:
Add both dones back in

The done functions were lost in the refactors. Let us add them back in.

danieljbruce committed:
mockedBeginTransaction should be Function

Make a more specific type for mockedBeginTransaction and the functions mocked.

danieljbruce committed:
Use ECMA script modifier

Make sure this function is completely hidden.

danieljbruce committed:
Remove TODOs that no longer apply

The TODO statements no longer need to be followed up on so remove them.

danieljbruce committed:
beginTransaction type definition

Remove the type definition. It is not used.

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-datastore into nodejs-transaction-redesign-feature-branch-2aa-adding-the-tests

danieljbruce committed:
Merge branch 'nodejs-transaction-redesign-feature-branch' of https://github.com/danieljbruce/nodejs-datastore into nodejs-transaction-redesign-feature-branch-2aa-adding-the-tests

# Conflicts:
#	src/transaction.ts
#	test/transaction.ts

danieljbruce committed:
Add the setupBeginTransaction method

Refactor test code to provide better error catching and also not repeat setup code for the run function.

danieljbruce committed:
Add a comment for private runCommit method

The comment for the runCommit method just captures the fact that the function is used as a pass-through.

danieljbruce committed:
Modify comment to be more clear

danieljbruce committed:
Update comments

The comments should more directly address the problem the tests are meant to solve

danieljbruce committed:
Eliminate redundant test code

This describe block is now duplicated. Remove it.

danieljbruce committed:
refactor: Move commit logic and add tests that prepare for transaction function changes

danieljbruce committed:
revert comment

This was a typo

danieljbruce committed:
Merge branch 'nodejs-transaction-redesign-feature-branch-2aa-adding-the-tests' of https://github.com/danieljbruce/nodejs-datastore into nodejs-transaction-redesign-feature-branch-2aa-adding-the-tests

danieljbruce committed:
Merge branch 'nodejs-transaction-redesign-feature-branch-2aa-adding-the-tests' of https://github.com/danieljbruce/nodejs-datastore into nodejs-transaction-redesign-feature-branch-5-adding-the-tests

# Conflicts:
#	src/transaction.ts
#	test/transaction.ts

danieljbruce committed:
runQuery, put, commit

Another some more integration tests for runQuery, put, commit

danieljbruce committed:
run linter and group some tests into a describe block so that they don’t run before some new tests we are going to add.

danieljbruce committed:
Add some runAggregationQuery integration tests

Add one of the test cases from the document (runAggregationQuery tests)

danieljbruce committed:
Add tests for put, runAggregationQuery, commit

put, runAggregationQuery, commit tests have been added and now run properly.

danieljbruce committed:
Write some latency tests

The latency tests measure time taken with and without using the run call.

danieljbruce committed:
Add logs to make latency tests run

The logs will output the time required for the latency tests.

danieljbruce committed:
Add two tests for put, commit

The two tests for put, commit should make sure that begin transaction is called.

danieljbruce committed:
Build requests into the transaction order tester

To meet the needs we want for unit testing we must add some checker that lets us record the requests and verify that they are the right values.

danieljbruce committed:
Modify order testing

Modify the order tester object to include more calls. Also add expected requests to the existing tests.

danieljbruce committed:
Create lookup, lookup, put, commit

Create test for a read, read and then commit.

danieljbruce committed:
Fix the unit test

Unit test should capture the fact that the run callback is used.

danieljbruce committed:
Add lookup requests

When using get requests without passing consistency, we should see two lookup requests reach the gapic layer. Consistency should be removed from the test because it is not meant to reach the Gapic layer.

danieljbruce committed:
Remove the console logs

console logs are not needed. Remove them so that the tests are cleaner.

danieljbruce committed:
Remove only

danieljbruce committed:
finish merge

danieljbruce committed:
remove the latency tests

The latency tests are not needed anymore so remove them.

danieljbruce committed:
Separate into a section transactions with/without

A section name transactions with and without run should be used.

danieljbruce committed:
Remove a describe block and run linter

This should make the diff a lot cleaner

danieljbruce committed:
Re-introduce a test that was there from before

We don’t want the diff to say that we have removed tests. Therefore, add the test that was there before.

danieljbruce committed:
Update comment

Modify the comment so that it is more objective

danieljbruce committed:
Add a comment for the runAggregationQuery function

The comment for the aggregation query function in transaction.ts is needed here.

danieljbruce committed:
Simplify argument type

In #withBeginTransaction there is a data type that matches the complex data type in the second argument of this function.

danieljbruce committed:
remove unnecessary extra variable

The extra variable is not needed

danieljbruce committed:
Eliminate redundant definition

Eliminate a definition and use the commit response data type here instead.

danieljbruce committed:
Add a comment for the #withBeginTransaction fn

Need to clearly explain what the withBeginTransactionFunction does

danieljbruce committed:
Eliminate a line that was used for debugging

danieljbruce committed:
Add space back in

danieljbruce committed:
remove indent

danieljbruce committed:
Add empty line back in

This simplifies the diff a bit

danieljbruce committed:
Remove ambiguous pronoun

danieljbruce committed:
Shorten comment

danieljbruce committed:
Add a comment for begin

Add a comment that incorporates how this function interacts with #withBeginTransaction.

danieljbruce committed:
Shorten comments

Don’t repeat information from the comments in the super class. Just mention the super class comments.

danieljbruce committed:
Type is only used once

Inline the type to eliminate redundancy in the code.

danieljbruce committed:
Inline another promise type

Inline another promise type which lets us eliminate one line of code.

danieljbruce committed:
Rename type to executor

Executor is what it is called in the promise argument.

danieljbruce committed:
run linter

danieljbruce committed:
Simplify the data types more

The types should line up for the executor in the runAsync function

danieljbruce committed:
Rename type to describe how it is used

This name is clearer

danieljbruce committed:
Rename to userCallbackData. Add comments

UserCallbackData is a more specific description of the type of data the generic type is. We also need comments to explain what these interfaces should represent.

danieljbruce committed:
resolver is a better term to use here

danieljbruce committed:
Change the type in the comments

The type has updated since to be a more general type.

danieljbruce committed:
Flip if/else

Flip the if else. Improvements are that the if checks for the affirmative and the longer block comes first

danieljbruce committed:
run linter

danieljbruce committed:
Add brackets

Introduce brackets so that the word matching the parameter is more isolated.

danieljbruce commented on src/transaction.ts:
```diff
@@ -579,26 +717,39 @@ class Transaction extends DatastoreRequest {
+   * begin a transaction that completed successfully.
+   *
+   **/
+  #parseRunSuccess(response: UserCallbackData<any>) {
```
We could change the response type here to `UserCallbackData<google.datastore.v1.IBeginTransactionResponse>`,  but then id would need to accept the `Buffer` type which would be a breaking change since the user can directly read `id`.

danieljbruce committed:
Stronger type checking for runAsync

Use more specific types in the runAsync function so that the compiler will complain if anything isn’t right.

danieljbruce committed:
Call the function fed into the promise resolver

resolver should be the name used just to be consistent with all other code.

danieljbruce committed:
Rename variables and add comments

Rename the variables so that the relationships between them are easier to understand. Add comments explaining the purpose of the variables.

danieljbruce committed:
Add comment for the done function

danieljbruce committed:
Make checkForCompletion private

To avoid confusion make checkForCompletion private so that the user of the object does not need to think about using it.

danieljbruce committed:
Replace all events with enum values

The events should not be strings. If they are strings then we might be magic string matching and this should not be done in code.

danieljbruce committed:
Make code more succinct

Remove the need for a map. Remove occurrences of strings throughout tests and replace them with enum values.

danieljbruce committed:
Rename event to UserCodeEvent

Call this enum UserCodeEvent as it is a better description of what the values could potentially be.

danieljbruce committed:
Remove the map from string names to gapic layer

This map is not needed anymore because now we use the Gapic layer function event directly.

danieljbruce committed:
More specific type

Shouldn’t just be a string. This should be a Gapic Layer function.

danieljbruce committed:
todo is done

Assertion check is done

danieljbruce committed:
Eliminate need to define extra variable

This just makes the code a bit more concise.

danieljbruce committed:
Add comments to describe test objects

The comments describe Gapic layer data passed back.

danieljbruce committed:
There is no need to make these functions private

Just inline the callbacks. A private function makes their relationship with the code that uses them confusing and each one of them is only used in one place.

danieljbruce committed:
Renames all callbacks to just callback

Rename all callbacks to callback because that is the variable name inside the client function.

danieljbruce committed:
Take requests private

danieljbruce committed:
Take expectedRequests private

expectedRequests is not used outside of the transaction order tester so we should take it private.

danieljbruce committed:
Take expected event order private

Expected event order should be taken private.

danieljbruce committed:
Take event order private

event order is not used outside this object

danieljbruce committed:
Fix some warning messages

Various warnings show up in the es-lint. Fix them.

danieljbruce committed:
Use unknown for Gapic Layer Response

Change the type to unknown to address the ES-lint error.

danieljbruce committed:
Get rid of any type for runQuery

Use the more specific type provided by runQuery instead.

danieljbruce committed:
More linter fixes

Addressing some of the problems in the new tests according to ES-lint.

danieljbruce committed:
Change parameters to match the get callback

Stronger type enforcement helping us catch errors later.

danieljbruce committed:
Another ESLInt correction

Use GetCallback here instead of any for the Get response type.

danieljbruce committed:
Remove query variable as it is not used

danieljbruce committed:
Take done private

When done is passed in and saved it should only be used internally.

danieljbruce committed:
Eliminate unused arguments in callback

Eliminate the arguments because if they are not used then they cause ES-lint errors.

danieljbruce committed:
Add readOnly reserved word

Various properties can be readOnly as recommended by the linter.

danieljbruce committed:
Make type for done explicit

Explicitly call out the fact that done should equal mocha’s done function.

danieljbruce committed:
Another place where done should be mentioned

The done type should explicitly be mentioned here.

danieljbruce committed:
No need to access order tester to get wrapper

There is no need to access the transaction wrapper to get the order tester. Just access the transaction wrapper directly.

danieljbruce committed:
private variable transaction wrapper

Take transaction wrapper private. It does not need to be visible to the user.

danieljbruce committed:
Capitalize RequestType

Change RequestType to be capitalized and use it whenever a request needs to be made.

danieljbruce committed:
request should be a more specific type

RequestType should be used here.

danieljbruce committed:
Do info checks for error and response cases

Add another assertion check. Do the check for the error and response cases.

danieljbruce committed:
These are actually results from runQuery

danieljbruce committed:
Eliminate extra line of code

Add assertion check for info. Eliminate the code that stores all the results.

danieljbruce committed:
rename enum

The word Layer is redundant. Also emphasize that the enum captures the function name.

danieljbruce committed:
Lookup response

The variable should include the word lookup since technically that is the name of the Gapic layer endpoint.

danieljbruce committed:
Add comments for the transaction order tester

Explain each function with a comment describing what it does.

danieljbruce committed:
Introduce a new push function

Add a push function so that in each test it is easier to see the events line up.

danieljbruce committed:
Remove middle layer functions

Remove functions in the transactionOrderTester and use functionality from those functions directly because then in the test it is easier to read the test’s intent. Also, add GapicRequestData type to simplify types in the test class. Shorten transactionOrderTester to just tester.

danieljbruce committed:
Prefer block comments

The comments are multi-line so use block comments instead.

danieljbruce committed:
Rename event to CUSTOM_EVENT

CUSTOM_EVENT should be used instead of FUNCTIONS_CALLED. FUNCTIONS_CALLED is sort of ambiguous.

danieljbruce committed:
Add square brackets to comments

square brackets added to make the comment more readable and user friendly.

danieljbruce committed:
Remove unused comment

danieljbruce committed:
Eliminate the promiseType variable

This variable is only used once. We should not define it here.

cindy-peng commented on src/transaction.ts:
```diff
@@ -161,7 +197,53 @@ class Transaction extends DatastoreRequest {
+  ): Promise<UserCallbackData<T>> {
+    if (this.#state === TransactionState.NOT_STARTED) {
+      const release = await this.#mutex.acquire();
+      try {
```
Nitpick: is it possible to make this a "try {} catch {} finally {}" block instead of using nested try blocks?

danieljbruce replied:
Yup. Done.

danieljbruce committed:
Eliminate double try/catch blocks

The nested try/catch blocks are not necessary. The finally block is going to run even if it returns an error in the catch block.

danieljbruce committed:
Add comments to describe what T is

Need to make sure the generic type is clear.

danieljbruce commented on src/transaction.ts:
```diff
@@ -22,26 +22,60 @@ import {google} from '../protos/protos';
+interface UserCallbackData<T> {
   err?: Error | null;
-  resp: any; // TODO: Replace with google.datastore.v1.IBeginTransactionResponse and address downstream issues
+  resp?: T;
```
Done

danieljbruce commented on src/transaction.ts:
```diff
@@ -161,7 +197,53 @@ class Transaction extends DatastoreRequest {
+          release();
+        }
+      } catch (err: any) {
+        return {err};
```
I initially thought the finally would not run with the earlier return, but it looks like this isn't the case. This has been flattened now.

danieljbruce commented on src/transaction.ts:
```diff
@@ -161,7 +197,53 @@ class Transaction extends DatastoreRequest {
+   *
+   * @param {CallOptions | undefined} [gaxOptions]
+   * @param {Resolver<T>} [resolver]
+   * @returns {Promise<UserCallbackData<T>>} Returns a promise that will run
```
In `#withBeginTransaction`, `{err}` is `UserCallbackData<T>` type and `await new Promise(resolver)` is also of type `UserCallbackData<T>`. Note that `return await new Promise(resolver)` doesn't actually return a promise because the `await` gets the value that the promise resolves to and then the return returns that value. Since the function is `async`, the value provided after `return` is automatically wrapped in a promise and the promise is returned to caller of this function.

Note that when this function runs the transaction and then runs the resolver code, like a transaction get or commit for example, then only the results from the resolver code will be returned to the user. If we delivered the `RunResponse` data back to the user then the signatures of get, runQuery, runAggregationQuery and commit would have to change. The user doesn't have to know about the `RunResponse` data anyway. If the transaction run call results in an error however then the error will be delivered to the user.

danieljbruce commented on src/transaction.ts:
```diff
@@ -161,7 +197,53 @@ class Transaction extends DatastoreRequest {
+        } finally {
+          release();
+        }
+      } catch (err: any) {
```
It looks like if it is anything else then we get an error that says `Catch clause variable type annotation must be 'any' or 'unknown' if specified.`. Throughout the codebase normally we just have `} catch (err) {` so maybe I will just adjust to match that.

danieljbruce replied:
But if I remove `: any` then I get a `Type 'unknown' is not assignable to type 'Error | null | undefined'.` so I think I am going to leave this as is unless you have a specific suggestion.

daniel-sanche replied:
I'm not a node person so I can't say, it just stood out to me. I'd expect we'd want more specific typing, but maybe that can't be done here

danieljbruce committed:
Use wrapped promise

The wrapped promise and runExclusive should be used instead of releasing the mutex manually. This is safer.

danieljbruce committed:
Change name to callback instead

Name the variable to match the parameter name being passed into runExclusive.

danieljbruce committed:
Eliminate extra variable

Use the async function in place. Don’t define an extra variable just for one usage.

danieljbruce committed:
Add a test for commit first

If commit is first then make sure the calls happen in the right order.

danieljbruce committed:
Remove a redundant check for transaction started

The results won’t be any different if the first check is removed. Therefore, should remove it just because it makes the code a lot simpler.

danieljbruce committed:
Rename variable and use runExclusive

runExclusive in the run function will get rid of excess code that is not needed. Also rename a variable to runResults to be more specific.

danieljbruce commented on src/transaction.ts:
```diff
@@ -161,7 +197,53 @@ class Transaction extends DatastoreRequest {
+    resolver: Resolver<T>
+  ): Promise<UserCallbackData<T>> {
+    if (this.#state === TransactionState.NOT_STARTED) {
+      const release = await this.#mutex.acquire();
```
Yes. `runExclusive` will be used now instead of `acquire`.

danieljbruce commented on src/transaction.ts:
```diff
@@ -446,9 +571,31 @@ class Transaction extends DatastoreRequest {
+              response: UserCallbackData<google.datastore.v1.IBeginTransactionResponse>
+            ) => {
+              release();
+              this.#processBeginResults(response, callback);
```
https://github.com/googleapis/nodejs-datastore/pull/1205/files#diff-7aa7c28abfaf34d94e1f9755555bd2571bad57fb652aad577c041cd5adf630c6R1018 is a test that captures this case and ensures that the `beginTransaction` call only happens once. The placement of release doesn't matter because the release is an async event that gets added to the end of the event loop.

In any case, I replaced this with `runExclusive` so that we don't have to trigger the release programmatically.

danieljbruce commented on src/transaction.ts:
```diff
@@ -579,26 +726,39 @@ class Transaction extends DatastoreRequest {
+  #parseRunSuccess(response: UserCallbackData<any>) {
+    const resp = response.resp;
+    this.id = resp!.transaction;
+    this.#state = TransactionState.IN_PROGRESS;
```
No. Unfortunately the id in code has a dual purpose. Normally id stores the transaction id returned from beginTransaction. It can also be used to specify the retry transaction id in the first `beginTransaction` call by supplying id in the constructor and saving it at https://github.com/googleapis/nodejs-datastore/blob/75949b51ea37f2e316a0818b969796a98ce2ba10/src/transaction.ts#L77. If the user supplies the id in the constructor then we still want `state === NOT_STARTED` so we can't rely on `id` to set `state` to `NOT_STARTED` or `IN_PROGRESS`.

danieljbruce commented on src/transaction.ts:
```diff
@@ -161,7 +197,53 @@ class Transaction extends DatastoreRequest {
+
+  /**
+   * If the transaction has not begun yet then this function ensures the transaction
+   * has started before running the resolver provided. The resolver is a function with one
```
There is a bigger discussion to be had here. First, before we add a decorator, let's make the changes from https://github.com/googleapis/nodejs-datastore/pull/1206 so that it is easier to identify the right pattern. I see some ways the decorator can be added, but they add too many levels of indirection. I'll think about it some more though. In the mean time I think the changes from https://github.com/googleapis/nodejs-datastore/pull/1206 should be added.

daniel-sanche replied:
That's fine if decorators don't fit, just wondering if it was considered

daniel-sanche commented on src/transaction.ts:
```diff
@@ -22,26 +22,60 @@ import {google} from '../protos/protos';
+ */
+interface Resolver<T> {
+  (resolve: PromiseResolveFunction<T>): void;
+}
```
These types are all private, right?

danieljbruce replied:
Yup. To make them available outside this module you would have to add `export` before the word `interface`.

danieljbruce committed:
refactor: remove resolvers, remove promise in withBeginTransaction, simplify code calling withBeginTransaction (#1212)

* Change resolvers

Change the resolvers so that they all follow the same pattern.

* Add a wrapWithBeginTransaction function for get

Generalize the calls to withBeginTransaction. Don’t allow the error type to be undefined in UserCallbackData since the error never actually will be undefined. This is necessary to solve various compile time errors.

* Use wrapWithBeginTransaction in more places

commit and runQuery can now use #wrapWithBeginTransaction and this will reduce the amount of code necessary in each function respectively.

* Add callbackWithError function

This function replaces a common pattern that occurs in transactions where we provide a callback that accepts an error as the first argument and some generic type that extends an array of any as the rest of the arguments.

* Add return type to callbackWithError

Adding a return type just makes type checking stronger in the client library. We should introduce it here to make it more likely a compiler error will be thrown if the function is misused.

* Delete code that is not used

When playing around with code to return a standard callback, code was written that is not required anymore. Also, rename StandardCallbackArgumentsAny.

* There is no need to define a separate type

The UserCallbackArguments do not need a separate type if they are only used once. Define the type inline where it is needed.

* TODO no longer applies

* Remove TODO

* Inline the error type

Error or null is only used in two places so inline it here.

* Add a comment for callbackWithError

callbackWithError is an abstract function that needs to be explained. This comment helps with that.

* Added comments to document sendUserCallbackData

sendUserCallbackData needs to be described so that its purpose is clear.

* Move function

Move function to the right order alphabetically

* Do not define a commit type only used once

Do not define a separate variable for only one use.

* Replace the type in the comment with Function

Replace with Function to eliminate references to the generic parameter which make CI tests fail.

* Just use function type for the callback

No need to mention the argument types and return types.

* Remove withBeginTransaction

withBeginTransaction is only called by one function so remove it to condense call stack.

* Simplify code usage of withBeginTransaction

Eliminate the need to build a resolver for all code usages of withBeginTransaction. Eliminate the need for a promise in withBeginTransaction. Eliminate the need for a callbackWithError function.

* function should not have any arguments

* Remove generic Args parameter

It is not used.

* Remove TODOs that are not relevant anymore.

* Rename to withBeginTransaction

Rename the function and eliminate redundant code.

* Update comments for withBeginTransaction

Update the comments to reflect new parameters.

* Simplify diff

* Remove error as null

* withBeginTransaction

Move withBeginTransaction to last alphabetically

* Add a comment to indicate error

In the code block it is not clear what kind of error will be produced and caught in catch so a comment is added to clarify the type of error that gets produced.

* Rename method to beginTxAsync

Calling the method `run` is too ambiguous. Use #beginTransactionAsync instead.

daniel-sanche commented on src/transaction.ts:
```diff
@@ -22,26 +22,57 @@ import {google} from '../protos/protos';
+ */
+interface Resolver<T> {
+  (resolve: PromiseResolveFunction<T>): void;
+}
```
This is my attempt to understand what these types represent. Let me know if I am thinking about these right

```
**
 *  Some methods in the Transaction class allow the user to pass in arbitrary call-backs
 * UserCallBackData holds the response from the passed-in callback, along with an error if encountered
 */
interface UserCallbackData<T>
  
// T represents the eventual result data from the user's callback
resp?: T;

/**
 * This function will execute the callback passed in, even if it is wrapped in a promise
 */
interface PromiseResolveFunction<T>

/**
 * A Resolver function does some work, and then calls PromiseResolveFunction to finalize the user-supplied callback (if any)
 */
interface Resolver<T> 
```

daniel-sanche commented on src/transaction.ts:
```diff
@@ -838,6 +1020,44 @@ class Transaction extends DatastoreRequest {
+              const runResults = await this.#beginTransactionAsync({
+                gaxOptions,
+              });
+              this.#parseRunSuccess(runResults);
```
Comments would be helpful here. Maybe at 1047, say that you are doing an rpc call to get the transaction id, then at 1051 say that the rpc was successful and you'll now fall through to run the requested `fn`

daniel-sanche commented on src/transaction.ts:
```diff
@@ -838,6 +1020,44 @@ class Transaction extends DatastoreRequest {
+
+  /**
+   * If the transaction has not begun yet then this function ensures the transaction
+   * has started before running the function provided as a parameter.
```
I'd probably re-word this, something like: 

```
Some rpc require that the transaction has been started (i.e, has a valid id) before they can be sent. #withBeginTransaction acts as a wrapper over those functions. 

If the function has not begun yet, `#withBeginTransaction` will first send an rpc to begin the transaction, and then execute the wrapped function. If it has begun, the wrapped function will be called directly instead

If an error is encountered during the transaciton begin call, the callback will be executed instead of the wrapped function
```

daniel-sanche commented on src/transaction.ts:
```diff
@@ -645,7 +747,87 @@ class Transaction extends DatastoreRequest {
         }
       );
     };
-    return new Promise(promiseFunction);
```
```
  (err, resp) => {
    resolve({
      err,
      resp,
    });
  }
```

Is the function expression really needed here? I though resolve was already a function. Couldn't this block just be replaced by `resolve`?

danieljbruce committed:
Remove abstract types and inline instead

With changes from before, these abstract data types are now only used once in the async function that begins a transaction. This change inlines the code that creates the promise so that these don’t have to be used at all.

danieljbruce committed:
Change the type of the id

The id for a transaction should actually match the value returned by the server. Adding this change also allows us to be more specific with types in function arguments.

danieljbruce committed:
For types to align, previous txn must be flexible

Previous transaction must accept the Uint8Array type because after all, that is what the id type usually is.

danieljbruce committed:
Set transaction type in request options

The request options must accept a Uint8Array for the compiler to work now that the ids can accept this type too.

danieljbruce committed:
For read options, more types should be accepted

The type for read options should line up with the type for other properties that store a transaction id.

danieljbruce committed:
Replace UserCallbackData with BeginAsyncResponse

UserCallbackData is only used in one place. Therefore, the generic type is not necessary and this can be replaced with a more specific type for readability.

danieljbruce committed:
Rewrite description for withBeginTransaction.

This description adds a few more details.

danieljbruce committed:
Add comments to withBeginTransaction

Ensure that the use of withBeginTransaction is clear and that this function contains comments throughout it.

danieljbruce committed:
Make the MockedTransactionWrapper class flexible

We want to test what happens when sending an error back from the begin transaction call.

danieljbruce committed:
Move the after hook out

The after hook does the same thing for all four of the describe blocks. Move it out to reduce the amount of code required.

danieljbruce committed:
Add beginTransaction error test

Add a test to ensure that when beginTransaction sends back an error that the error actually reaches the user when they are using a promise or a callback.

danieljbruce committed:
Add comments to withBeginTransaction

Comments are needed in withBeginTransaction to clearly indicate the flow of control within the function.

danieljbruce reviewed (commented):
Addressed latest feedback. Flattened data structures and removed abstractions in favour of a flatter call stack.

daniel-sanche commented on src/transaction.ts:
```diff
@@ -300,6 +322,46 @@ class Transaction extends DatastoreRequest {
+      () => {
+        super.get(keys, options, callback);
+      },
+      callback
```
nit: some comments inside the function could be helpful

daniel-sanche reviewed (approved):
That last change was a huge improvement! LGTM as-is, although some more comments in the functions still might make it easier for future maintainers

danieljbruce committed:
Added a few comments for withBeginTxn

Comments will help all functions that are using withBeginTransaction.
